File: services/k8s-service-go/internal/routes/custom_resources.go
package routes

import (
	"github.com/go-chi/chi/v5"

	"github.com/junginho0901/kubeast/services/k8s-service-go/internal/handler"
)

// RegisterCustomResources — CRDs and custom resource instances.
func RegisterCustomResources(r chi.Router, h *handler.Handler) {
	// Custom Resource Definitions (cluster-scoped)
	r.Get("/api/v1/crds", h.GetCRDs)
	r.Get("/api/v1/crds/{name}/describe", h.DescribeCRD)
	r.Delete("/api/v1/crds/{name}", h.DeleteCRD)

	// Custom Resource Instances
	r.Get("/api/v1/custom-resources/all", h.GetAllCustomResourceInstances)
	r.Get("/api/v1/custom-resources/{group}/{version}/{plural}", h.GetCustomResourceInstances)
	r.Get("/api/v1/custom-resources/{group}/{version}/{plural}/{namespace}/{name}/describe", h.DescribeCustomResourceInstance)
	r.Delete("/api/v1/custom-resources/{group}/{version}/{plural}/{namespace}/{name}", h.DeleteCustomResourceInstance)
}

File: frontend/src/services/api/custom_resources.ts
// Custom Resource API — CRDs (CustomResourceDefinitions) and the
// generic instance lookup that dispatches by group/version/plural.

import { client } from './client'
import type { CRDInfo, CustomResourceInstanceInfo } from './types'

export const customResourcesApi = {
  // CRDs
  getCRDs: async (forceRefresh = false): Promise<CRDInfo[]> => {
    const { data } = await client.get('/cluster/crds', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  describeCRD: async (name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/crds/${name}/describe`)
    return data
  },

  deleteCRD: async (name: string): Promise<void> => {
    await client.delete(`/cluster/crds/${name}`)
  },

  // Instances
  getAllCustomResourceInstances: async (forceRefresh = false): Promise<CustomResourceInstanceInfo[]> => {
    const { data } = await client.get('/cluster/custom-resources/all', {
      params: { force_refresh: forceRefresh },
    })
    return data
  },

  getCustomResourceInstances: async (group: string, version: string, plural: string): Promise<any[]> => {
    const { data } = await client.get(`/cluster/custom-resources/${group}/${version}/${plural}`)
    return data
  },

  describeCustomResourceInstance: async (group: string, version: string, plural: string, namespace: string, name: string): Promise<any> => {
    const { data } = await client.get(`/cluster/custom-resources/${group}/${version}/${plural}/${namespace}/${name}/describe`)
    return data
  },

  deleteCustomResourceInstance: async (group: string, version: string, plural: string, namespace: string, name: string): Promise<void> => {
    await client.delete(`/cluster/custom-resources/${group}/${version}/${plural}/${namespace}/${name}`)
  },
}
